go/networkqualityd.go

```diff
@@ -91,17 +104,54 @@ func main() {
 		}()
 	}
 
+	if *announce {
+		ips := make([]net.IP, 0)
+		// The user may give us a hostname (rather than an address to listen on). In order to
+		// handle this situation, we will use DNS to convert it to an IP. As a result, we may
+		// get back more than one address -- handle that!
+		if addresses, lookupErr := net.LookupHost(*listenAddr); lookupErr == nil {
+			for _, addr := range addresses {
+				if parsedAddr := net.ParseIP(addr); parsedAddr != nil {
+					ips = append(ips, parsedAddr)
+				}
+			}
+		}
+		if announceResponder, announceHandle, announceErr := configureAnnouncer(ips, *configName, *configPort); announceErr == nil {
+			defer announceResponder.Remove(announceHandle)
+			go announceResponder.Respond(operatingCtx)
+		} else {
+			log.Printf("Warning: Could not announce the server instance: %v.\n", announceErr)
+		}
+	}
+
 	var wg sync.WaitGroup
 	wg.Add(1)
 
-	listenAddr := fmt.Sprintf("%s:%d", *listenAddr, *configPort)
-	go func(listenAddr string) {
-		log.Printf("Network Quality URL: https://%s:%d%s/config", *configName, *configPort, *contextPath)
-		if err := http.ListenAndServeTLS(listenAddr, *certFilename, *keyFilename, mux); err != nil {
+	server := &http.Server{}
+	server.Addr = fmt.Sprintf("%s:%d", *listenAddr, *configPort)
+	server.Handler = mux
+
+	go func(server *http.Server, configName string, configPort int, configContextPath string) {
+		log.Printf("Network Quality URL: https://%s:%d%s/config", configName, configPort, configContextPath)
+		if err := server.ListenAndServeTLS(*certFilename, *keyFilename); err != nil && err != http.ErrServerClosed {
 			log.Fatal(err)
 		}
 		wg.Done()
-	}(listenAddr)
+	}(server, *configName, *configPort, *contextPath)
```

Comment:
I was reading up on Go's internal architecture today and came across this gem:

> To avoid mistakes, only read-only values should be left to be implicitly captured by closures invoked in goroutines; this makes sense from a concurrency standpoint as well. [[1]](https://eli.thegreenplace.net/2019/go-internals-capturing-loop-variables-in-closures/)

So, even though they are read-only variables in the closure, I figured that we could make things explicit rather than implicit.